### ISSUE TITLE
[HOLD for payment 2023-09-07] [$4000] Inconsistent front-end email validation across the product

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17387]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
(Chat) 
1. Create a new chat
2. Enter an invalid email (e.g. jaj@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfisjksksjsjssskssjskskssksksksksskdkddkddkdksskskdkdkdksskskskdkdkdkdkekeekdkddenejeodxkdndekkdjddkeemdjxkdenendkdjddekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdi.cdjd
3. Start the chat

(Login/SignUp)
1. Try to login with an invalid email like jaj@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfisjksksjsjssskssjskskssksksksksskdkddkddkdksskskdkdkdksskskskdkdkdkdkekeekdkddenejeodxkdndekkdjddkeemdjxkdenendkdjddekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdi.cdjd

(Add Contact Method) 
1. Navigate to Settings > Profile > Contact Method > New Contact Method 
4. Try to add an invalid email like jaj@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfisjksksjsjssskssjskskssksksksksskdkddkddkdksskskdkdkdksskskskdkdkdkdkekeekdkddenejeodxkdndekkdjddkeemdjxkdenendkdjddekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdi.cdjd

## Expected result: 
Front-end validation should prevent these emails from being added and creating unnecessary network requests. Additionally, consistent validation should be used in all areas where we validate emails in the product (e.g. pasting an invalid email into an existing chat still formats it like a valid email)

Backend email validation uses [PHP's email validation filter](https://www.php.net/manual/en/filter.filters.validate.php) FILTER_VALIDATE_EMAIL as well as the following regex (for emails and phone numbers) [\w\.'#%+-]+@[a-zA-Z\d\.-]+\.[a-zA-Z]{2,}|^\+?[1-9]\d{1,14}$)

## Actual results: 
Emails are successfully submitted, only to be subsequently rejected by the backend

## Workaround:
N/A

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** v1.3.0-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/231782900-50ff85fb-d63d-481a-b754-6265979f3f68.mp4

<img width="960" alt="Untitled" src="https://user-images.githubusercontent.com/43996225/231782943-79319911-6a0d-4bc2-83e6-f2713cae2030.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @chiragxarora
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681358100970919

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01878e05ad15a14027</li>
        <li>Upwork Job ID: 1648410114624286720</li>
        <li>Last Price Increase: 2023-08-10</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Reviewer | 26133961</li>
<li>jjcoffee | Contributor | 26133964</li>
</ul></li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
@adelekennedy Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by adelekennedy** [comment]
I can reproduce! Seems very similar to #17388

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01878e05ad15a14027

**Comment by MelvinBot** [comment]
Current assignee @adelekennedy is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @aimane-chnaif (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NikkiWines (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by jjcoffee** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
As the [Slack convo](https://expensify.slack.com/archives/C049HHMV9SM/p1681358100970919) indicates, and I think what we really care about here, is that there is an inconsistency between how the New Chat screen validates emails as you're typing and how we decide when to convert an "email-like" string to a link in the composer. IMO these two should at a minimum behave the same to reduce inconsistencies.

### What is the root cause of that problem?
We are using two different sets of regex patterns, one that ends up being used when validating emails as we type when starting a new chat, which can be traced back to `EMAIL_BASE_REGEX`:

https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/CONST.jsx#L3

and the markdown one here, which eventually is used to converts emails to links:

https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/CONST.jsx#L333

This duplicating of regexes appears to be a result of [this PR](https://github.com/Expensify/expensify-common/pull/508). I've asked there for clarification on why two are needed, as the justification seems odd (that the original one is too old).

### What changes do you think we should make in order to solve the problem?
I think we should either merge the two rules or choose the one we want to use everywhere (unless there's a really good reason to have separate ones). We can then easily modify and keep behaviour consistent in different parts of the app.

I think the priority would be to add back in the validation from `EMAIL_BASE_REGEX` (i.e. applied when searching when starting a New Chat), that validates the TLD is at least 2 characters long. 

I'm not sure if I interpret this comment correctly, but it seems that `EMAIL_BASE_REGEX` (referenced via `REG_EXP.EMAIL`) regex is kept in sync with the backend, so it might be the best to use.

https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/str.js#L389-L397

If we do want to implement a length limit we could add it to the existing`EMAIL_BASE_REGEX` so that it becomes:

`((?=.{1,254}$)[\\w\\-\\+\\'#]+(?:\\.[\\w\\-\\'\\+]+)*@(?:[\\w\\-]+\\.)+[a-z]{2,})`

Or if we decide the `MARKDOWN_EMAIL` rule is better, then that would become:

```
((?=.{1,254}$)[a-zA-Z0-9.!#$%&'+/=?^`{|}-][a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]*@[a-zA-Z0-9-]+?(\\.[a-zA-Z]+)+)
```

Personally I think a merging of the two rules is in order, as `MARKDOWN_EMAIL` handles more characters before the @ (such as \`), whereas `EMAIL_BASE_REGEX` works better for the domain and TLD (e.g. handles multiple dashes, like `test@staging0-new.base-site.com`), so something like:

```
((?=.{1,254}$)[a-zA-Z0-9.!#$%&'+/=?^`{|}-][a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]*@(?:[\\w\\-]+\\.)+[a-z]{2,})
```

FWIW I don't know how useful it is to impose an email length limit here as we're not really in the business of fully validating the email (for example we still accept incorrect [TLDs](https://data.iana.org/TLD/tlds-alpha-by-domain.txt), like `.example`), we just want to turn any reasonably email-looking string into a link. If someone wants to deliberately type an incredibly long email, they're free to do it - as long as we're not expecting it to validate as a real email on the BE. (The same is true of the [related issue](https://github.com/Expensify/App/issues/17388) that was closed).

**Comment by chiragxarora** [comment]
Hi @jjcoffee ,
You're right about the duplicate regex rules and those two can be combined into one which is being used in the backend (if it is being used).
I believe probability of a user coming in and typing a very long email like string is very low but having a limit on length makes our app more consistent with any modern day chat application in the market. Again possibility of someone entering a thousands of character long email is very low but if we render it as a link which opens in an email app, that would reflect as overall bad aesthetics and validation. Since we cannot accurately validate the TLDs without the DNS lookup, the least we can do is validation on length part on frontend.

**Comment by abdulrahuman5196** [comment]
I am coming from the PR which introduced MARKDOWN_EMAIL regex.

I am not sure if understand the problem statement correctly or what is trying to be fixed here.  

1) The problem statement expects "should not parse the email address as a valid one for abc@abc.abc". 
2) And the proposal and slack conversation is talking about limiting email length to RFC standard. 
3) And we are also checking to make common regex for emails(this was also talked in the original PR which introduced MARKDOWN_EMAIL).

I am confused here, Are we trying to solve all 3 above or something specifically? Could you kindly provide this information so I can provide better information with regards to the original MARKDOWN_EMAIL regex. @aimane-chnaif

**Comment by chiragxarora** [comment]
Sure @aimane-chnaif 

Jaj@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjeJajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfisjksksjsjssskssjskskssksksksksskdkddkddkdksskskdkdkdksskskskdkdkdkdkekeekdkddenejeodxkdndekkdjddkeemdjxkdenendkdjddekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjeJajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdi.cdjd

**Comment by NikkiWines** [comment]
Sorry for the delay here! Couple things looking through this issue. 

> Enter any invalid email having atleast 3 parts [abc@abc.abc](https://www.expensify.com/_support/?supportEmail=abc%40abc.abc&comment=better%20supportal%20issue%20nubmer%3A%2017387&location=_support/)(can extend from here)

`.abc` (and `.om` from the testing screenshots) are both valid TLDs, so some of the emails in the OP that we're saying are invalid are actually valid. `.ddddd` and `.cdjd` don't seem to be valid though. 

> FWIW I don't know how useful it is to impose an email length limit here as we're not really in the business of fully validating the email (for example we still accept incorrect [TLDs](https://data.iana.org/TLD/tlds-alpha-by-domain.txt), like .example), we just want to turn any reasonably email-looking string into a link. If someone wants to deliberately type an incredibly long email, they're free to do it - as long as we're not expecting it to validate as a real email on the BE. (The same is true of the https://github.com/Expensify/App/issues/17388 that was closed).

Yep, I agree with all of this and can confirm it's not successfully passing on the BE 

> @NikkiWines can you please share all email validation regexes used in backend along with where they're used?

From some local testing I can tell you that it's failing [PHP's email validation filter](https://www.php.net/manual/en/filter.filters.validate.php) `FILTER_VALIDATE_EMAIL`. Also I don't think you need the usage, as the parsing for new Expensify should be handled in [Expensimark.js](https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js).

**Comment by chiragxarora** [comment]
Hi @NikkiWines 
correct me if I'm wrong but I think validation on domain names including TLD's is not really possible since any organization can register for their custom domains and they get added constantly. What we were trying to do was adding a validation check on length of the email address because that's what possible on frontend. To accurately validate a domain name I don't think we have any option other than dns lookup..

**Comment by NikkiWines** [comment]
> correct me if I'm wrong but I think validation on domain names including TLD's is not really possible since any organization can register for their custom domains and they get added constantly. 

No, you're correct, and I'm not suggesting we move to validating emails based on TLDs - just mentioning that the emails used in the OP aren't actually invalid (at the very least, the `abc@abc.abc` shouldn't be)

**Comment by NikkiWines** [comment]
In this case, it doesn't seem that there's a unified standard for handle these extra long emails, I think our current formatting (which seems to be the same as Githubs right now) is fine. @adelekennedy, @aimane-chnaif thoughts?

 Here are some other examples: 

- Gmail
<img width="609" alt="Screenshot 2023-04-24 at 16 26 04" src="https://user-images.githubusercontent.com/16747822/234136836-f5bfddd4-e543-40c4-82b1-e86478bab7d5.png">

- Slack
<img width="485" alt="Screenshot 2023-04-24 at 16 24 25" src="https://user-images.githubusercontent.com/16747822/234136829-664596de-29a2-4beb-b8a0-0900937c1125.png">

- Whatsapp 
<img width="456" alt="Screenshot 2023-04-24 at 16 25 06" src="https://user-images.githubusercontent.com/16747822/234136833-1d34c452-e720-41c5-b27c-5314efee6e75.png">

- Github 

jaj@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfisjksksjsjssskssjskskssksksksksskdkddkddkdksskskdkdkdksskskskdkdkdkdkekeekdkddenejeodxkdndekkdjddkeemdjxkdenendkdjddekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdi.cdjd

abc@abc.abc

jajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfisjksksjsjssskssjskskssksksksksskdkddkddkdksskskdkdkdksskskskdkdkdkdkekeekdkddenejeodxkdndekkdjddkeemdjxkdenendkdjddekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejdicdjejajasjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjcdidjjcdkekdiccjdkejdjcjxisdjjdkedncicdjejejcckdsijcjdsodjcicdkejd@gmail.com

**Comment by adelekennedy** [comment]
@NikkiWines that seems fine to me - there doesn't really seem to be a standard we're conforming to, and as the user I don't really feel it as a pain point tbh

**Comment by jjcoffee** [comment]
There's also inconsistency between different parts of the frontend, e.g. searching for a user when starting a new chat uses a different validation from the markdown one. I think it makes sense to maximise consistency where possible.

**Comment by NikkiWines** [comment]
> Right now it's inconsistent between frontend and backend. Those kinds of long emails are acceptable in frontend but not in backend. This will break flow when add contact in offline mode.

> searching for a user when starting a new chat uses a different validation from the markdown one. I think it makes sense to maximise consistency where possible.

Yeah, good points @aimane-chnaif  and @jjcoffee, these are both compelling reasons to modify our current front-end validation logic. I agree it would be good to be consistent. 

Aside from the PHP validation mentioned earlier, we also use the following regex to validate emails and phone numbers `[\w\.'#%+-]+@[a-zA-Z\d\.-]+\.[a-zA-Z]{2,}|^\+?[1-9]\d{1,14}$)`

**Comment by MelvinBot** [comment]
@NikkiWines @adelekennedy @aimane-chnaif this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by chiragxarora** [comment]
`hh`

**Comment by NikkiWines** [comment]
I actually think we should probably make a new GitHub issue for this since the current issue has changed significantly from the OP. It will also make it easier for us to track who should receive what payment and account for the timeline of the bug and proposal reviewing process. 

Thoughts @adelekennedy? If we're agreed, this should be raised in the #expensify-bugs slack channel as a new `BUG:` post with proper reproduction steps

**Comment by chiragxarora** [comment]
Shouldn't the original bug reporter be the same for the bug report since everything revolves around the main invalid email validation issue?

**Comment by alexrasla** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

In this problem, we are able to add and initiate a chat with email ids that are invalid.

### What is the root cause of that problem?

The root cause of the problem is the fact that the `EMAIL_BASE_REGEX` regular expression constant in `node_modules/expensify-common/lib/CONST.jsx` does not check for character limits for email ids and domains. The constant only checks that the compared string has valid email characters and an @ sign through regular expressions.

### What changes do you think we should make in order to solve the problem?


The change I would make to solve this is simply changing the `EMAIL_BASE_REGEX` to check if the email id has the correct type and number of characters in the regular expression. This would produce the `Invalid email` error on any page as shown below so the user would never be able to add the email address.

![Screenshot 2023-05-04 at 9 24 39 AM](https://user-images.githubusercontent.com/38263534/236266913-5219c1d5-00f4-4387-ac61-628f6248f38b.png)

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by MelvinBot** [comment]
@NikkiWines @adelekennedy @aimane-chnaif this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by NikkiWines** [comment]
Chatted with some other internal employees and decided to just update this issue + double the price to try and get things moving. cc: @aimane-chnaif

**Comment by jjcoffee** [comment]
@aimane-chnaif I've updated my [proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314) with a more fully fledged regex to match the BE. 

I think it might be good if we could agree on a set of test emails to run through (I have some on my proposal, but open to adding more!).

**Comment by NikkiWines** [comment]
@aimane-chnaif, please review the updated proposals above, thank you! 🙇

**Comment by kushu7** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Inconsistent front-end email validation across the product
### What is the root cause of that problem?
We don't have matching validation as per backend.
### What changes do you think we should make in order to solve the problem?

We can use this regex in front-end. it is compatible with backend php `FILTER_VALIDATE_EMAIL`.


```js
^(?!(?:(?:\x22?\x5C[\x00-\x7E]\x22?)|(?:\x22?[^\x5C\x22]\x22?)){255,})(?!(?:(?:\x22?\x5C[\x00-\x7E]\x22?)|(?:\x22?[^\x5C\x22]\x22?)){65,}@)(?:(?:[\x21\x23-\x27\x2A\x2B\x2D\x2F-\x39\x3D\x3F\x5E-\x7E]+)|(?:\x22(?:[\x01-\x08\x0B\x0C\x0E-\x1F\x21\x23-\x5B\x5D-\x7F]|(?:\x5C[\x00-\x7F]))*\x22))(?:\.(?:(?:[\x21\x23-\x27\x2A\x2B\x2D\x2F-\x39\x3D\x3F\x5E-\x7E]+)|(?:\x22(?:[\x01-\x08\x0B\x0C\x0E-\x1F\x21\x23-\x5B\x5D-\x7F]|(?:\x5C[\x00-\x7F]))*\x22)))*@(?:(?:(?!.*[^.]{64,})(?:(?:(?:xn--)?[a-zA-Z0-9]+(?:-+[a-zA-Z0-9]+)*\.){1,126}){1,}(?:(?:[a-zA-Z][a-zA-Z0-9]*)|(?:(?:xn--)[a-zA-Z0-9]+))(?:-+[a-zA-Z0-9]+)*)|(?:\[(?:(?:IPv6:(?:(?:[a-fA-F0-9]{1,4}(?::[a-fA-F0-9]{1,4}){7})|(?:(?!(?:.*[a-fA-F0-9][:\]]){7,})(?:[a-fA-F0-9]{1,4}(?::[a-fA-F0-9]{1,4}){0,5})?::(?:[a-fA-F0-9]{1,4}(?::[a-fA-F0-9]{1,4}){0,5})?)))|(?:(?:IPv6:(?:(?:[a-fA-F0-9]{1,4}(?::[a-fA-F0-9]{1,4}){5}:)|(?:(?!(?:.*[a-fA-F0-9]:){5,})(?:[a-fA-F0-9]{1,4}(?::[a-fA-F0-9]{1,4}){0,3})?::(?:[a-fA-F0-9]{1,4}(?::[a-fA-F0-9]{1,4}){0,3}:)?)))?(?:(?:25[0-5])|(?:2[0-4][0-9])|(?:1[0-9]{2})|(?:[1-9]?[0-9]))(?:\.(?:(?:25[0-5])|(?:2[0-4][0-9])|(?:1[0-9]{2})|(?:[1-9]?[0-9]))){3}))\]))$

```

Result: 

https://github.com/Expensify/App/assets/31707153/b076bdad-dbe2-4e6b-9a3a-9715f661adb1




### What alternative solutions did you explore? (Optional)
None

**Comment by jjcoffee** [comment]
@aimane-chnaif I've made sure that the validation matches the BE, see the list of test cases in my [proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314), which have tested successfully against the existing BE validation.

The additional BE regex (from [here](https://github.com/Expensify/App/issues/17387#issuecomment-1522544213)) is more restrictive on which characters it accepts than the PHP `FILTER_VALIDATE_EMAIL` filter (e.g. it won't accept `email{@domain.com`, whereas PHP would), so we only really need the length limits from the PHP filter and then we can just use the BE regex for character validation. I have confirmed from testing that the additional BE regex is just applied on top of the PHP filter, so it should be perfectly aligned now.

**Comment by jjcoffee** [comment]
@aimane-chnaif I've popped a test script up here if you want to double-confirm: https://onlinephp.io/c/d2aad

**Comment by NikkiWines** [comment]
> I see there are still 2 different validations in backend. Can you please specify use cases for each one so that frontend does the same?

We use both validations in the back end for processing any/all emails/phone numbers, there aren't specific use cases for one vs the other. I think the best approach would be to try and make a combination of both validations for the front-end validation.

**Comment by adelekennedy** [comment]
catching up here, are we looking for an amended proposal? Should this issue be external or remain internal?

**Comment by adelekennedy** [comment]
should we update this to add the external label @aimane-chnaif or do you want to review the existing proposals and then add the external label if needed?

**Comment by jjcoffee** [comment]
@aimane-chnaif I'm curious, is there anything wrong with my [proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314)? It tests perfectly for me, but maybe I'm missing something!

**Comment by payal-lathidadiya** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Inconsistent front-end email validation across the product. Validation passed in front-end gets failed from backend which causing inconsistent behavior.

### What is the root cause of that problem?
The current implementation of `isValidEmail` in front-end, only checks that the string parameter has valid email characters and an @ sign through regular expressions. This is not sufficient and differing from backend validation.

### What changes do you think we should make in order to solve the problem?

We should modify `isValidEmail` function to check individual parts length along with regular expression check which will check the format.

Update regular express as;
```
const EMAIL_BASE_REGEX = /^[-!#$%&'*+\/0-9=?A-Z^_a-z`{|}~](\.?[-!#$%&'*+\/0-9=?A-Z^_a-z`{|}~])*@[a-zA-Z0-9](-*\.?[a-zA-Z0-9])*\.[a-zA-Z](-?[a-zA-Z0-9])+$/;
```

Update validation function with;
```
isValidEmail(string) {
  if (!string) return false;

  var emailParts = string.split('@');

  if (emailParts.length !== 2) return false;

  var account = emailParts[0];
  var address = emailParts[1];

  if (account.length > 64) return false;

  else if (address.length > 255) return false;

  var domainParts = address.split('.');
  
  if (domainParts.some(function (part) {
    return part.length > 63;
  })) return false;

  return Boolean(String(string).match(CONST.REG_EXP.EMAIL));
};
``` 

**References:**
- http://fightingforalostcause.net/misc/2006/compare-email-regex.php
- http://thedailywtf.com/Articles/Validating_Email_Addresses.aspx
- http://stackoverflow.com/questions/201323/what-is-the-best-regular-expression-for-validating-email-addresses/201378#201378
- https://en.wikipedia.org/wiki/Email_address  

The format of an email address is local-part@domain, where the local part may be up to 64 octets long and the domain may have a maximum of 255 octets.

### What alternative solutions did you explore? (Optional)
Or we can create deep email validator solution to cover below checks and have our custom solution;
- To validate common typos e.g. [example@gmaill.com](mailto:example@gmaill.com) using [mailcheck](https://github.com/mailcheck/mailcheck).
- To validate email was not generated by disposable email service using [disposable-email-domains](https://github.com/ivolo/disposable-email-domains).
- To validate MX records are present on DNS.
- To Validate SMTP server is running.
- To Validate mailbox exists on SMTP server.

**Comment by NikkiWines** [comment]
@aimane-chnaif any update on the proposals above?

**Comment by NikkiWines** [comment]
@aimane-chnaif friendly bump here 🙇

**Comment by payal-lathidadiya** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/17387#issuecomment-1554420680) 

### What changes do you think we should make in order to solve the problem?

From the research I've done so far, it's said to not to use or rely only on regular expression for email validation due to extreme complexity.


<details>
  <summary>references</summary>

-   [Please do NOT use regex to “validate” email addresses](https://michaellong.medium.com/please-do-not-use-regex-to-validate-email-addresses-e90f14898c18#:~:text=Share-,Please%20do%20NOT%20use%20regex%20to%20%E2%80%9Cvalidate%E2%80%9D%20email%20addresses.,a%20perfectly%20valid%20email%20address)
- [3 reasons why you should stop using regex email validation](https://www.loqate.com/en-gb/blog/3-reasons-why-you-should-stop-using-regex-email-validation/)
- [How to verify that strings are in valid email format](https://learn.microsoft.com/en-us/dotnet/standard/base-types/how-to-verify-that-strings-are-in-valid-email-format)

</details>

However, we could use one of the following standard regEXs to check for the valid structure of an email.
1. RFC 5322 compliant regex
   ```
   (?:[a-z0-9!#$%&'*+=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])
    ```
1. RFC822 (way complicated approach - A Perl module has a [long regular expression](http://ex-parrot.com/~pdw/Mail-RFC822-Address.html) based on the standard description of an email address)
    [RegEx](http://www.ex-parrot.com/~pdw/Mail-RFC822-Address.html)

1. Simple and widely used
    ```
    /^[A-Za-z0-9_!#$%&'*+\/=?`{|}~^.-]+@[A-Za-z0-9.-]+$/gm
    ```

We can combine one of this format check along with length validation suggested in previous proposal for different individual parts.

**Comment by jjcoffee** [comment]
@aimane-chnaif It would be helpful if you could elaborate on what's not perfect. Are there some test cases that you're using that I'm missing on my [proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314) that are failing?

**Comment by adelekennedy** [comment]
@JmillsExpensify reassigning to myself after ooo

**Comment by NikkiWines** [comment]
Looks like we're still waiting on an appropriate proposal here

**Comment by adelekennedy** [comment]
Still pending proposals!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by jeet-dhandha** [comment]
I am working on this give me like 24 hours 👍 . Will be creating a unified regex for email and markdown which we can use for both.

**Comment by adelekennedy** [comment]
If this issue isn't reproducible should we still be looking for proposals @aimane-chnaif?

**Comment by adelekennedy** [comment]
Let's keep it going then @jeet-dhandha we'll look forward to your proposal

**Comment by jeet-dhandha** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
- We want a common email validation which should be consistent with markdown and email checks.

### What is the root cause of that problem?
- We have implemented two distinct sets of regex patterns `EMAIL_BASE_REGEX` and `MARKDOWN_EMAIL`.

### What changes do you think we should make in order to solve the problem?
- I have created a common funciton using the markdown (as it covers every rule) regex and added a `_` for email validation as emails might have starting as `_`.

### Updated Code:
<img width="1308" alt="Screenshot 2023-06-14 at 10 43 13 AM" src="https://github.com/Expensify/App/assets/78416198/35f5f70d-7526-49eb-b9b0-64ffe310fe39">

### Extra Test Cases

<img width="1309" alt="Screenshot 2023-06-14 at 10 55 20 AM" src="https://github.com/Expensify/App/assets/78416198/4bacc8fb-ec04-4fcf-9da2-8583163d5154">
<img width="938" alt="Screenshot 2023-06-14 at 10 55 53 AM" src="https://github.com/Expensify/App/assets/78416198/4a93f41a-c2e8-4794-bb7e-daa173d1c9e9">

### Test Results:
<img width="561" alt="Screenshot 2023-06-14 at 10 56 44 AM" src="https://github.com/Expensify/App/assets/78416198/8c507cd5-8fdd-4319-842a-8483d780fa10">

**Comment by jjcoffee** [comment]
Proposal [updated](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314).

One thing to clarify is how we'd want to handle the conflict between emails starting with an underscore and markdown italics (since `_example@test.com` is a valid email according to the BE regex). 

It's easy enough to handle simple cases like `_example@test.com_` (just match the email within the underscores), but it gets more complex for multi-line mixed examples, e.g.

```
_email@test.com is my main email, you can also try _email2@test.com

or:
_email3@test.com_
```

where it becomes less clear which parts should be italicised.

The easiest solution is just not to accept leading underscores in markdown since it's really rare to actually see an email starting with an underscore, but I don't know if that would be acceptable @aimane-chnaif?

**Comment by jjcoffee** [comment]
@adelekennedy Is it possible for this one to be doubled up? It's a pretty large amount of work that affects a lot of areas of the app, not to mention that it's been open for 2 months by now! Maybe @aimane-chnaif wants to weigh in too.

**Comment by adelekennedy** [comment]
Yep - doubling this one

**Comment by securitybeautystar** [comment]
### **Proposal:**
Understanding of Project Goals:

Your product currently lacks robust front-end email validation, resulting in unnecessary network requests when users enter invalid email addresses. You want a consistent and efficient validation approach that aligns with your existing back-end PHP validation, preventing invalid emails from being submitted.

Proposal and Implementation Approach:

Front-end Validation: I will implement a JavaScript-based validation function that matches the PHP's FILTER_VALIDATE_EMAIL functionality and your existing regex. This function will run each time an email address is input, ensuring its validity before it is submitted.

Unified Validation Approach: This validation will be applied across all sections of the product where email input is required – including creating a new chat, login/signup, and adding a new contact method. This ensures a consistent experience for the users.

Error Messaging: To make the application more user-friendly, I will include clear and informative error messages for instances where the entered email address does not pass the validation checks.

Testing: Rigorous testing will be performed to ensure the validation works as expected, and no invalid emails are passed through to the back end.

**Comment by jeet-dhandha** [comment]
@adelekennedy https://github.com/Expensify/App/issues/17387#issuecomment-1590493021 Any updates on my proposal ?

**Comment by adelekennedy** [comment]
@aimane-chnaif when you get a chance will you review the updated proposal above?

**Comment by jjcoffee** [comment]
Proposal [updated](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314) (again!) with additional markdown tests. A few more issues cropped up when running those new tests, so I've also tweaked the regex to handle the full range of cases, including the italic markdown conflict mentioned previously. All tests are now passing and validate well against the BE. cc @aimane-chnaif.

**Comment by jjcoffee** [comment]
@Louis46cp Ah thanks for pointing that out, I hadn't added that part to my proposal. I'm using the base regex everywhere, updated now!

**Comment by jjcoffee** [comment]
@Louis46cp True, but I think in that case it's fairly obvious what it's meant to refer to. I'm trying to walk the line between not providing a full diff (which is against the guidelines) and providing workable code/general approach, without creating a massive wall of text of a proposal 😅 [Updated](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314) anyway!

**Comment by jeet-dhandha** [comment]
Cool 👍  - Should i create a PR, as i am not allowed to create branch on `expensify-common`

**Comment by jjcoffee** [comment]
@aimane-chnaif Sure, test branch is [here](https://github.com/jjcoffee/expensify-common/tree/fix-17387). You can use [this](https://onlinephp.io/c/f31bd) PHP script to validate tests against BE extra regex + PHP's `FILTER_VALIDATE_EMAIL`.

**Comment by jjcoffee** [comment]
[Proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314) and [branch](https://github.com/jjcoffee/expensify-common/tree/fix-17387) updated with additional tests from @YacineF0001 and a tweak to the regex. cc @aimane-chnaif

**Comment by payal-lathidadiya** [comment]
## Proposal

New thoughts with follow up to my below previous proposals
- https://github.com/Expensify/App/issues/17387#issuecomment-1554420680
- https://github.com/Expensify/App/issues/17387#issuecomment-1564703595
I could see there are lots of proposals being posted here suggesting different regex to be used which will cover more scenarios. As per my research and analysis, I found no regex can cover 100% use cases being email validation is very complex.
I think we should focus on the core problem statement for which we faced issues;
- Inconsistency in front-end and back-end validation.
And to resolve this we should have same validation logic between frontend and backend so that it won’t cause inconsistency wherein client side validated request pass through front-end and then failed by backend validation logic.

### What changes do you think we should make in order to solve the problem?

My proposal here is to use same regular expression for frontend which is used by PHP’s backend `FILTER_VALIDATE_EMAIL`
ref: https://github.com/php/php-src/blob/PHP-8.2.6/ext/filter/logical_filters.c#L682
```js
// the regex I have used is from PHP version 8.2.6 which is used in php_filter_validate_email
const EMAIL_BASE_REGEX = ‘/^(?!(?:(?:\\x22?\\x5C[\\x00-\\x7E]\\x22?)|(?:\\x22?[^\\x5C\\x22]\\x22?)){255,})(?!(?:(?:\\x22?\\x5C[\\x00-\\x7E]\\x22?)|(?:\\x22?[^\\x5C\\x22]\\x22?)){65,}@)(?:(?:[\\x21\\x23-\\x27\\x2A\\x2B\\x2D\\x2F-\\x39\\x3D\\x3F\\x5E-\\x7E\\pL\\pN]+)|(?:\\x22(?:[\\x01-\\x08\\x0B\\x0C\\x0E-\\x1F\\x21\\x23-\\x5B\\x5D-\\x7F\\pL\\pN]|(?:\\x5C[\\x00-\\x7F]))*\\x22))(?:\\.(?:(?:[\\x21\\x23-\\x27\\x2A\\x2B\\x2D\\x2F-\\x39\\x3D\\x3F\\x5E-\\x7E\\pL\\pN]+)|(?:\\x22(?:[\\x01-\\x08\\x0B\\x0C\\x0E-\\x1F\\x21\\x23-\\x5B\\x5D-\\x7F\\pL\\pN]|(?:\\x5C[\\x00-\\x7F]))*\\x22)))*@(?:(?:(?!.*[^.]{64,})(?:(?:(?:xn--)?[a-z0-9]+(?:-+[a-z0-9]+)*\\.){1,126}){1,}(?:(?:[a-z][a-z0-9]*)|(?:(?:xn--)[a-z0-9]+))(?:-+[a-z0-9]+)*)|(?:\\[(?:(?:IPv6:(?:(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){7})|(?:(?!(?:.*[a-f0-9][:\\]]){7,})(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,5})?::(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,5})?)))|(?:(?:IPv6:(?:(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){5}:)|(?:(?!(?:.*[a-f0-9]:){5,})(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,3})?::(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,3}:)?)))?(?:(?:25[0-5])|(?:2[0-4][0-9])|(?:1[0-9]{2})|(?:[1-9]?[0-9]))(?:\\.(?:(?:25[0-5])|(?:2[0-4][0-9])|(?:1[0-9]{2})|(?:[1-9]?[0-9]))){3}))\\]))$/iDu’;
```

### What alternative solutions did you explore? (Optional)
None

**Comment by adelekennedy** [comment]
pending proposal reviews

**Comment by jjcoffee** [comment]
[Proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314) and [branch](https://github.com/jjcoffee/expensify-common/tree/fix-17387) updated with additional tweaks and tests for a few more edge cases I've come across. cc @aimane-chnaif

**Comment by adelekennedy** [comment]
pending updated proposal reviews @aimane-chnaif

**Comment by jeet-dhandha** [comment]
@aimane-chnaif 
> `jeet-dhandha:fixed-markdown` - Branch for https://github.com/Expensify/App/issues/17387#issuecomment-1590493021

Did you checked this branch ?

**Comment by adelekennedy** [comment]
@aimane-chnaif @NikkiWines are we ready to move forward with @jjcoffee proposal? Is there anything else that needs to be discussed/updated here?

**Comment by NikkiWines** [comment]
Hmm, no, nothing appears to have changed in the email validation regex we use - should still be the following `[\w\.'#%+-]+@[a-zA-Z\d\.-]+\.[a-zA-Z]{2,}|^\+?[1-9]\d{1,14}` and [PHP's `FILTER_VALIDATE_EMAIL`](https://www.php.net/manual/en/filter.filters.validate.php)

@aimane-chnaif do you have an update on your last [comment](https://github.com/Expensify/App/issues/17387#issuecomment-1620733092)? Are we able to move forward here?

**Comment by jjcoffee** [comment]
@chiragxarora The local part of that email (before the @) is longer than 64 chars, so wouldn't have ever passed BE validation. The overall length can be longer (e.g. the domain part being repeated chunks of 63 chars), so that might be what you're thinking of!

**Comment by Antasel** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue.
at Frontend, validating some emails which is not valid at backend side.

### What changes do you think we should make in order to solve the problem?
we have inconsistency in Frontend email validation(different regex patterns are used for EMAIL_BASE_REGEX and MARKDOWN_EMAIL) and those are not matched to backend's email validation

### What changes do you think we should make in order to solve the problem?

**1. First, We can get regex for FILTER_VALIDATE_EMAIL(referenced from PHP open source code) from [https://github.com/php/php-src/blob/PHP-8.2.6/ext/filter/logical_filters.c#L682](https://github.com/php/php-src/blob/PHP-8.2.6/ext/filter/logical_filters.c#L682)**
```
/^(?!(?:(?:\\x22?\\x5C[\\x00-\\x7E]\\x22?)|(?:\\x22?[^\\x5C\\x22]\\x22?)){255,})(?!(?:(?:\\x22?\\x5C[\\x00-\\x7E]\\x22?)|(?:\\x22?[^\\x5C\\x22]\\x22?)){65,}@)(?:(?:[\\x21\\x23-\\x27\\x2A\\x2B\\x2D\\x2F-\\x39\\x3D\\x3F\\x5E-\\x7E\\pL\\pN]+)|(?:\\x22(?:[\\x01-\\x08\\x0B\\x0C\\x0E-\\x1F\\x21\\x23-\\x5B\\x5D-\\x7F\\pL\\pN]|(?:\\x5C[\\x00-\\x7F]))*\\x22))(?:\\.(?:(?:[\\x21\\x23-\\x27\\x2A\\x2B\\x2D\\x2F-\\x39\\x3D\\x3F\\x5E-\\x7E\\pL\\pN]+)|(?:\\x22(?:[\\x01-\\x08\\x0B\\x0C\\x0E-\\x1F\\x21\\x23-\\x5B\\x5D-\\x7F\\pL\\pN]|(?:\\x5C[\\x00-\\x7F]))*\\x22)))*@(?:(?:(?!.*[^.]{64,})(?:(?:(?:xn--)?[a-z0-9]+(?:-+[a-z0-9]+)*\\.){1,126}){1,}(?:(?:[a-z][a-z0-9]*)|(?:(?:xn--)[a-z0-9]+))(?:-+[a-z0-9]+)*)|(?:\\[(?:(?:IPv6:(?:(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){7})|(?:(?!(?:.*[a-f0-9][:\\]]){7,})(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,5})?::(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,5})?)))|(?:(?:IPv6:(?:(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){5}:)|(?:(?!(?:.*[a-f0-9]:){5,})(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,3})?::(?:[a-f0-9]{1,4}(?::[a-f0-9]{1,4}){0,3}:)?)))?(?:(?:25[0-5])|(?:2[0-4][0-9])|(?:1[0-9]{2})|(?:[1-9]?[0-9]))(?:\\.(?:(?:25[0-5])|(?:2[0-4][0-9])|(?:1[0-9]{2})|(?:[1-9]?[0-9]))){3}))\\]))$/iDu
```
This pattern consists of main Email definition and length limitation for email ID(<65), domain's each level after TLD(<64), total length(<255), TLD length(1< tld length <64).

**2. we get the unified regex with applying backend's additional regex(`[\w\.'#%+-]+@[a-zA-Z\d\.-]+\.[a-zA-Z]{2,}`) to step 1's regex**
EMAIL_BASE_REGEX: 
```
`(?=((?![^@\\s]{65})[\\w'#%+-]+(?:\\.(?:[\\w'#%+-]+))*@(?:(?![^.\\s]{64})[a-z0-9]+(?:-+[a-z0-9]+)*\\.)+[a-z]{2,63}(?= |_|\\b))(?<end>.*))\\S{3,254}(?=\\k<end>$)`
```

MARKDOWN_EMAIL is equal to EMAIL_BASE_REGEX

The autoEmail rule's Regex needs to be updated to this:

`(?![^<]*>|[^<>]*<\\/(?!em))(?:(?<= |[^\\w'#%+-])|^|\\b)${CONST.REG_EXP.MARKDOWN_EMAIL}(?!((?:(?!<a).)+)?<\/a>|[^<]*(<\/pre>|<\/code>))`

**3. We have another issue in how to apply italic and autoEmail rule into email started with underscore( _ ).**
Simple solution is that BE side does not validate Email started with underscore.

If not, another solution is that we apply italic rule ,and then autoEmail rule.
to do that, move autoEmail, quote rule after italic rule and add a tweak for email started with underscore on italic rule's replacement

![image](https://github.com/Expensify/App/assets/139204153/95a03594-42c4-4485-8de2-6efcbad04e1d)

- with my approach, all of @jjcoffee 's cases are passed successfully.
 and added a edge case: 

string : 
`$--test@gmail.com`
result:
`$<a href=\"mailto:--test@gmail.com\">--test@gmail.com</a>`

![image](https://github.com/Expensify/App/assets/139204153/8c363f71-0964-4d2e-8cb9-01c8eac62071)

**4. Please check [this branch](https://github.com/Antasel/expensify-common/tree/fix-email-validation-inconsistency)**

**Comment by Antasel** [comment]
@jjcoffee 's solution does not cover following edge case.

test string : 
`$--test@gmail.com`

correct result:
`$<a href=\"mailto:--test@gmail.com\">--test@gmail.com</a>`

your result:
`$--<a href=\"mailto:test@gmail.com\">test@gmail.com</a>`

**Comment by jjcoffee** [comment]
Slack doesn't generate @Antasel's proposed processing of that email (`$--test@gmail.com`):
![image](https://github.com/Expensify/App/assets/27287420/56ea652f-75d4-40d9-abc2-1d6f1464a25b)

Same with Gmail:

![image](https://github.com/Expensify/App/assets/27287420/0ca57d86-fb4a-4c34-8e4e-a498744880df)

And WhatsApp:

![image](https://github.com/Expensify/App/assets/27287420/91102d46-9497-442d-9f1b-ceefa8dd4da9)

**Comment by Antasel** [comment]
@aimane-chnaif 
please check php [script](https://onlinephp.io/c/a4e66)
current backend rule allows it.

**Comment by jaytechfulldev** [comment]
@aimane-chnaif
did you check my propsal?

https://github.com/Expensify/App/issues/17387#issuecomment-1593907164

**Comment by jjcoffee** [comment]
To be clear `--test@example.com` is a valid email (which is handled correctly in my current proposal), but `$--test@example.com` is not. The question is regarding how to perform the markdown conversion in the latter case. Personally I don't think this case is a big deal as it's not really a valid string in the sense that the dollar sign doesn't mean anything syntactically so it's not so obvious what the intention is if someone actually typed that!

However, if we do want to handle it, it's an easy tweak to my proposal, i.e. the autoEmail rule can be tweaked to lookahead for any non-valid email chars, which obviates the need for the word boundary char: 
`(?![^<]*>|[^<>]*<\\/)(?:(?<= |[^\\w'#%+-])|^)${CONST.REG_EXP.MARKDOWN_EMAIL}(?![^<]*(<\\/pre>|<\\/code>|<\\/a>))`

I've now updated my [proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1514794314) and [repo](https://github.com/jjcoffee/expensify-common/tree/fix-17387) with this and the new test case.

Thinking about this more though, I wonder if there's an argument for **not** converting substrings like this, e.g. do we actually expect `$%*)|¬;/&test@example.com` to apply the markdown conversion on the email part of the string? It might make more sense to only allow certain characters before the email, i.e. `[ \\(@_*~]` (opening bracket, mentions, markdown chars). Any thoughts @aimane-chnaif?

**Comment by Antasel** [comment]
> Thinking about this more though, I wonder if there's an argument for **not** converting substrings like this, e.g. do we actually expect `$%*)|¬;/&test@example.com` to apply the markdown conversion on the email part of the string? It might make more sense to only allow certain characters before the email, i.e. `[ \\(@_*~]` (opening bracket, mentions, markdown chars). Any thoughts @aimane-chnaif?

@jjcoffee I can't agree on it.
what if user typed as following?

"I have sent 100$,--agc@gmail.com is my account email, please confirm'

**Comment by jjcoffee** [comment]
@Antasel I guess we could also include punctuation, but I agree it's probably more complicated than it's worth to limit the characters. It doesn't look like this is done on other apps (Slack or GMail) either.

**Comment by Antasel** [comment]
@jjcoffee 
on here this platform, it's done as what i am saying.
100$,--abc@gmail.com

on the other hand, Slack does not allow to start from underscore for email validation, so we should also follow it ? then why did you take it in consideration to avoid conflict with italic rule?

**Comment by Antasel** [comment]
> Hmm, no, nothing appears to have changed in the backend - should still be `[\w\.'#%+-]+@[a-zA-Z\d\.-]+\.[a-zA-Z]{2,}|^\+?[1-9]\d{1,14}` and [PHP's `FILTER_VALIDATE_EMAIL`](https://www.php.net/manual/en/filter.filters.validate.php)

According to @NikkiWines 's update, we should take it concern.
@jjcoffee you agreed that --abe@gmail.com is valid email, then it should be converted properly in any case, i can't understand why you try to exclude the cases ($,--test@gmailc.om)

**Comment by jjcoffee** [comment]
@Antasel I'm not disagreeing with you :slightly_smiling_face: I was saying that Slack and GMail **don't** apply any sort of character limitation like I suggested (I've also updated my proposal to handle the case you mention).

Underscores are currently taken to be valid by BE so we should remain consistent with it, I think. We're only looking at Slack and other apps to get a ballpark sense of how the markdown conversion should work within sane limits.

**Comment by adelekennedy** [comment]
woof, this is a tricky one - @aimane-chnaif is there more information we need or agreement on expected behavior before choosing a proposal? It sounds like we're making our own rules here a bit on email referencing other company choices (Slack & Google)

**Comment by adelekennedy** [comment]
>If we don't wanna change backend email validation (I believe so) based on expected behavior
agree - @NikkiWines double checking with you

**Comment by payal-lathidadiya** [comment]
@aimane-chnaif @NikkiWines

It looks like Antasel's proposal [here](https://github.com/Expensify/App/issues/17387#issuecomment-1630415664) is duplicate of what I already proposed [here](https://github.com/Expensify/App/issues/17387#issuecomment-1604380720) to use PHP's `FILTER_VALIDATE_EMAIL` backend logic?

**Comment by NikkiWines** [comment]
Whoops, sorry for the delay. I don't think we need to update the backend email validation, so let's continue to use it as our baseline for what emails are valid and invalid.

**Comment by adelekennedy** [comment]
ty @NikkiWines

**Comment by adelekennedy** [comment]
@aimane-chnaif do any of the proposals above confirm [work with this comment?](https://github.com/Expensify/App/issues/17387#issuecomment-1634677646) Or are we still waiting for a matching proposal

**Comment by Antasel** [comment]
> It seems Antasel's proposal [here](https://github.com/Expensify/App/issues/17387#issuecomment-1630415664) is duplicate of what I already proposed [here](https://github.com/Expensify/App/issues/17387#issuecomment-1604380720) to use PHP's `FILTER_VALIDATE_EMAIL` backend logic.

@payal-lathidadiya it's not duplicate of your proposal, please inspect into it again.
Moreover your regex pattern can't cover @aimane-chnaif 's [steps](https://github.com/Expensify/App/issues/17387#issuecomment-1567273370)

**Comment by Antasel** [comment]
@aimane-chnaif I think current my solution had already covered all cases from issues mentioned above. but i added the cases mentioned in the issues.

please check [php script](https://onlinephp.io/c/1d053).

Regarding to backend email validation tested on above script, added unit test cases as following.

<details>
  <summary>Test cases</summary>

```
expect(Str.isValidEmail('abc-@mail.com')).toBeTruthy();
expect(Str.isValidEmail('abc#def@mail.com')).toBeTruthy();
expect(Str.isValidEmail('ganache-tasacks.0z+randomlongtexttoreachsomecharactersneedsommore@icloud.com')).toBeFalsy();
expect(Str.isValidEmail('ganache-tacks.0z+randomlongtexttoreachsomecharactersneedsommore@icloud.com')).toBeTruthy();
expect(Str.isValidEmail('ganache-tacks.0z+randomlongtexttor3eachsomecharactersneedsomemo_@icloud.com')).toBeTruthy();
expect(Str.isValidEmail('ganache-tacks.0z+randomlongtexttoreachsomecharactersneedsomemo_@icloud.com')).toBeTruthy();
expect(Str.isValidEmail('kkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkk@gmail.com')).toBeFalsy();
expect(Str.isValidEmail('@.com')).toBeFalsy();
expect(Str.isValidEmail('jane.doe@gmail..com')).toBeFalsy();
expect(Str.isValidEmail('jane.doe@gmail.com@example.com')).toBeFalsy();
expect(Str.isValidEmail('jane.doe@com')).toBeFalsy();
expect(Str.isValidEmail('jane.doe@-gmail.com')).toBeFalsy();
expect(Str.isValidEmail('jone.due@-gmail.com')).toBeFalsy();
expect(Str.isValidEmail('Hello@gmail.com.com')).toBeTruthy();
expect(Str.isValidEmail('—@—.com')).toBeFalsy();
```

</details>

please check [this branch](https://github.com/Antasel/expensify-common/tree/fix-email-validation-inconsistency)

**Comment by payal-lathidadiya** [comment]
> > It seems Antasel's proposal [here](https://github.com/Expensify/App/issues/17387#issuecomment-1630415664) is duplicate of what I already proposed [here](https://github.com/Expensify/App/issues/17387#issuecomment-1604380720) to use PHP's `FILTER_VALIDATE_EMAIL` backend logic.
> 
> @payal-lathidadiya it's not duplicate of your proposal, please inspect into it again. Moreover your regex pattern can't cover @aimane-chnaif 's [steps](https://github.com/Expensify/App/issues/17387#issuecomment-1567273370)

I think we should differ to C+ and expensify team to review and decide on this.

**Comment by jjcoffee** [comment]
@aimane-chnaif I've updated [my branch](https://github.com/jjcoffee/expensify-common/tree/fix-17387) with a few extra tests from the issues you've linked, and temporarily added comments indicating which issue they're from so that it's easy to verify. I don't think there are actually any new types of emails in the issues you've linked that have not already been covered, however.

Some comments on each issue:

- https://github.com/Expensify/App/issues/16762 - I don't think there are any extra tests to extract from here, as it's a separate issue to handle regarding deeplinks
- https://github.com/Expensify/App/issues/17388 - this is **not relevant**, we are not fixing URL regex here
- https://github.com/Expensify/App/issues/17839 - examples added, but note that they evaluate as true according to BE. The linked issue expects RFC compliance which we are **not** implementing here - we are only matching BE behaviour.
- https://github.com/Expensify/App/issues/18181 - examples added, but I believe we should remove later for clarity as they don't test anything new (just the 64 char length limit that I've [already added](https://github.com/jjcoffee/expensify-common/blob/607acb1e77f68a96a6f0791560abbcff24f86919/__tests__/Str-test.js#L55-L56) in my tests).
- https://github.com/Expensify/App/issues/18855 - same as above, example added but should remove
- https://github.com/Expensify/App/issues/20105 - example added
- https://github.com/Expensify/App/issues/21110 - added but we already test for incorrect domain prefix (hyphen in this case)
- https://github.com/Expensify/App/issues/21230 - not invalid according to BE (`.com` is a valid subdomain)
- https://github.com/Expensify/App/issues/21753 - I couldn't locate any problematic emails to add as tests, but agree that the linked issue needs to be held for this one ideally
- https://github.com/Expensify/App/issues/22032 - added but don't think it's a useful test to keep (otherwise we'd just spam the test with all possible iterations of invalid characters!)

**Comment by adelekennedy** [comment]
@aimane-chnaif will you be able to provide feedback this week?

**Comment by adelekennedy** [comment]
pending feedback

**Comment by NikkiWines** [comment]
@aimane-chnaif do you have an update here?

**Comment by adelekennedy** [comment]
Just noting to remove the overdue - I don't think we want to increase the price on this any more, there's been a lot of back and forth but 4k is the ceiling

**Comment by jeet-dhandha** [comment]
This should be fixed when we fix this.
https://github.com/Expensify/App/issues/24226

**Comment by adelekennedy** [comment]
@jeet-dhandha huh - is the correct issue linked? It hasn't been triaged yet I'd assume we'd close that as a dupe of this issue?

**Comment by jeet-dhandha** [comment]
Yes ,obviously @aimane-chnaif 😂

**Comment by jeet-dhandha** [comment]
How will it be possibly to fix in reverse 🤷🏻 ?

**Comment by Antasel** [comment]
@aimane-chnaif 
did you check my [proposal](https://github.com/Expensify/App/issues/17387#issuecomment-1630415664) ? , it covers all cases and matches with Backend email validation .

**Comment by adelekennedy** [comment]
@aimane-chnaif I'm really leaning towards closing this as we've been dragging thus out for a while and it isn't affecting Echat's usability - it's a consistency problem

**Comment by adelekennedy** [comment]
Then let's prioritize this and make a decision today

**Comment by jjcoffee** [comment]
@aimane-chnaif Thanks for the review, that sounds fair to me! If @Antasel agrees, they can raise the initial PR and I can run the tests.

**Comment by payal-lathidadiya** [comment]
@aimane-chnaif it looks like we're moving ahead with solution I proposed here https://github.com/Expensify/App/issues/17387#issuecomment-1604380720 to have the PHP's FILTER_VALIDATE_EMAIL backend logic. To have consistency in FE and BE validation instead of introducing any new improved regex version which may fail to handle all different cases.

**Comment by payal-lathidadiya** [comment]
@aimane-chnaif it looks like we're moving ahead with solution I proposed here https://github.com/Expensify/App/issues/17387#issuecomment-1604380720 to have the PHP's FILTER_VALIDATE_EMAIL backend logic. To have consistency in FE and BE validation instead of introducing any new improved regex version which may fail to handle all different cases.

Further, I was actively part of investigation, discovery and exploring different solution around this issue and had proposed few proposals from time to time to communicate new updates on proposal to match with new things being discussed during reviews.

Let me know If I could be of any help to apply best possible fix to this long standing issue. Thank you.

**Comment by payal-lathidadiya** [comment]
> @payal-lathidadiya this issue requires not only regex fix but also markdown <-> html fixes. And I requested above but you didn't follow that. So I can say that you just found PHP validation regex and didn't put more efforts to make it fit to our app.

Correct. I completely agree with your point. It does require few additional changes in existing markdown rule due to side effect of new solution. I hadn't apply detailing in original proposal as I assumed that's something can be considered and dealt in actual implementation and PR once direction and approach gets a go ahead.

>  If you want your proposal to be accepted, please follow [#17387 (comment)]

Awh! Looks like somehow I missed [this](https://github.com/Expensify/App/issues/17387#issuecomment-1620733092) follow up message completely while I was out on vacation for few days around those dates.

Alright! no worries if that's something can be worked out with finalized proposal and contributor. 

It was just a genuine thought to offer possible help in case needed(with I'm being working closely and active part of solution contributions for few linked issues which are OnHold for this). and had no intention to argue or question on final review decision though 😄

**Comment by Antasel** [comment]
@aimane-chnaif Thanks for your review and suggestion.
I totally agree on equal shared bounty, since i can see @jjcoffee put much efforts to solve this issue for a long time.
I am new here, so My current concern is that I can get a committed PR history to join C+ team later, or not. through double work on a PR.

**Comment by NikkiWines** [comment]
Wooo, thank you for all the hard work! Agreed that the combo proposal looks good. 

As for the italicized and underscore-prefixed emails, I agree with what you said  [here](https://github.com/Expensify/App/issues/17387#issuecomment-1677064889)

> Their handling of italic cases are also fine but I prefer taking italic precedence over autoEmail which is @Antasel's proposal. Personally I don't care too much about this edge case where email starts with _ is also wrapped inside italic markdown.

Assigning all relevant parties 🚀

**Comment by adelekennedy** [comment]
Amazing! For payment are we going to split the 4k issue payment for the two contributors?

**Comment by jjcoffee** [comment]
@Antasel Let us know when you expect to be able to raise a PR. You can also reach out on Slack (I'm @Joel (Davies) there) if you need any help!

**Comment by Antasel** [comment]
Sorry for late reply
I was struggling to fix android building issue, for now it's built successfully.
and sent email to join slack channel. after then, i can get help from other devs for coming other issue

**Comment by Antasel** [comment]
> @Antasel please let us know in advance. If you won't be able to make PR today, @jjcoffee will do instead to push this forward.

I will make first expensify-common PR today

**Comment by Antasel** [comment]
@aimane-chnaif The PR https://github.com/Expensify/expensify-common/pull/566 for expensify-common is ready for review

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by adelekennedy** [comment]
pending review

**Comment by NikkiWines** [comment]
PRs are in review

**Comment by adelekennedy** [comment]
PRs are still n review

**Comment by Antasel** [comment]
@NikkiWines @adelekennedy 
Could you kindly let me know where I should apply on Upwork ? Looks like previous job link is invalid for now

**Comment by jjcoffee** [comment]
@Antasel Payment will be handled on 7th September (when the regression period ends - assuming no regressions are found). You'll either directly get an offer on Upwork or will need to apply to the job (the BZ @adelekennedy will create a new job if the old one is invalid).

**Comment by Antasel** [comment]
@JmillsExpensify @NikkiWines 
Could you help me to get expensify open-source channel invitation on Slack ?
I've sent email to contributors@expensify.com, but still no on my anayetrasel@yahoo.com .

**Comment by adelekennedy** [comment]
@Antasel @jjcoffee reposted the job [here](https://www.upwork.com/jobs/~01f34c92b173053348)!

**Comment by Antasel** [comment]
@adelekennedy 
I've applied to the job

**Comment by adelekennedy** [comment]
@Antasel do you mind reapplying to contributors one last time? I'll ping the person on inbox duty so they can approve you - once you've been added I'll process payment

**Comment by Antasel** [comment]
@adelekennedy I've sent again
![image](https://github.com/Expensify/App/assets/139204153/ab553fb0-960e-499a-87f3-173e96d52b86)

**Comment by jjcoffee** [comment]
@adelekennedy Thanks, applied to the job!

**Comment by EliteDevSolution** [comment]
Hello, Expensive Team

```
import React, { useState } from 'react';
import { View, TextInput, Button, Alert } from 'react-native';
import validator from 'validator';

const EmailValidationExample = () => {
  const [email, setEmail] = useState('');

  const handleInputChange = (text) => {
    setEmail(text);
  };

  const handleSubmit = () => {
    if (validator.isEmail(email)) {
      // Valid email, proceed with submission
      Alert.alert('Success', 'Email is valid');
      // Add code to send the email to the server here
    } else {
      Alert.alert('Error', 'Invalid email address');
    }
  };

  return (
    <View>
      <TextInput
        placeholder="Enter email"
        value={email}
        onChangeText={handleInputChange}
      />
      <Button title="Submit" onPress={handleSubmit} />
    </View>
  );
};

export default EmailValidationExample;
```

In this example, we're using the validator.js library to validate the email. Make sure to install it using npm install validator.

The handleInputChange function updates the email state as the user types. When the user presses the submit button, the handleSubmit function is called. It checks if the email is valid using validator.isEmail(email).

On a successful validation, you can then proceed to submit the email to the server. Make sure that server-side validation is also implemented to ensure that only valid emails are accepted.

Remember, this is a basic example and you may need to adapt it to fit your specific application structure and requirements. Additionally, you should handle edge cases and potential error scenarios appropriately.

For server-side validation, you can use PHP's FILTER_VALIDATE_EMAIL or the provided regex in your backend code to further ensure that only valid emails are accepted.

**Comment by EliteDevSolution** [comment]
Contributor details
Your Expensify account email: superstorm609@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~014918ff9ae7e9ef65

**Comment by EliteDevSolution** [comment]
Hello, Expensive Team
```
const validator = require('validator');

const phoneNumber = '+1334567890';

if (validator.isMobilePhone(phoneNumber, 'any')) {
  console.log('Valid phone number');
} else {
  console.log('Invalid phone number');
}
```
I think that no problem with phone number verification too.

**Comment by Antasel** [comment]
@adelekennedy 24 hours passed. Still didn't get invitation on my email. Could you please ping the inbox person again for me?
Thanks in advance

**Comment by adelekennedy** [comment]
<img width="1203" alt="Screenshot 2023-09-12 at 3 46 45 PM" src="https://github.com/Expensify/App/assets/60450566/70c7830c-5775-4bb4-ba23-daa554bbaf56">

@Antasel did you respond to Ben?

**Comment by EliteDevSolution** [comment]
@adelekennedy 24 hours passed. Still didn't get invitation on my email. Could you please ping the inbox person again for me?
Thanks in advance

**Comment by adelekennedy** [comment]
Have you replied to Ben? I added the screenshot of the email exchange above

**Comment by adelekennedy** [comment]
### Payouts due:

Issue Reporter: $250 @chiragxarora
Contributor: $2000 @jjcoffee & @Antasel 
Contributor+: $4000 @aimane-chnaif 

Eligible for 50% #urgency bonus? **N**

Upwork job is [here](https://www.upwork.com/jobs/~01f34c92b173053348).

**Comment by adelekennedy** [comment]
@chiragxarora will you [apply](https://www.upwork.com/jobs/~01f34c92b173053348) to the Upwork job for the reporting bonus?

**Comment by adelekennedy** [comment]
What's your Upwork information?

**Comment by adelekennedy** [comment]
<img width="504" alt="Screenshot 2023-09-14 at 11 30 47 AM" src="https://github.com/Expensify/App/assets/60450566/66e2141b-6bc0-44ba-8b9d-585c36c157c0">

**Comment by Antasel** [comment]
@chiragxarora please confirm your upwork information

**Comment by jjcoffee** [comment]
@adelekennedy Thanks, offer accepted!

**Comment by Antasel** [comment]
@adelekennedy accepted offer, thanks

**Comment by adelekennedy** [comment]
Offer sent @chiragxarora

**Comment by adelekennedy** [comment]
oh @aimane-chnaif I'm so sorry I thought you were in the NewDot payment beta as well, I'll update the summary above and send you an offer through Upwork
